server/src/main/java/org/elasticsearch/action/bulk/ConcurrentBulkProcessor.java

```diff
@@ -0,0 +1,153 @@
+    private final Runnable onClose;
+    private final Supplier<BulkRequest> bulkRequestSupplier;
+    private final AtomicInteger executionIdGen = new AtomicInteger();
+    private final AtomicReference<State> state = new AtomicReference<>(EMPTY_STATE);
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final Scheduler.Cancellable cancellableFlushTask;
+
+    public ConcurrentBulkProcessor(BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer, BackoffPolicy backoffPolicy,
+                                   BulkProcessor.Listener listener, int concurrentRequests, long bulkActions, ByteSizeValue bulkSize,
+                                   TimeValue flushInterval, Scheduler scheduler, Runnable onClose,
+                                   Supplier<BulkRequest> bulkRequestSupplier) {
+        this(consumer, backoffPolicy, listener, concurrentRequests, bulkActions, bulkSize, flushInterval, scheduler, scheduler, onClose,
+            bulkRequestSupplier);
+    }
+
+    public ConcurrentBulkProcessor(BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer, BackoffPolicy backoffPolicy,
+                                   BulkProcessor.Listener listener, int concurrentRequests, long bulkActions, ByteSizeValue bulkSize,
+                                   TimeValue flushInterval, Scheduler flushScheduler, Scheduler retryScheduler, Runnable onClose,
+                                   Supplier<BulkRequest> bulkRequestSupplier) {
+        this.bulkSize = bulkSize.getBytes();
+        this.bulkActions = bulkActions;
+        this.onClose = onClose;
+        this.bulkRequestSupplier = bulkRequestSupplier;
+        this.bulkRequestHandler = new ConcurrentBulkRequestHandler(concurrentRequests, consumer, backoffPolicy, retryScheduler, listener);
+        this.cancellableFlushTask = startFlushTask(flushInterval, flushScheduler);
+    }
+
+    public void add(IndexRequest request) {
+        updateState(request, BulkRequest.getSize(request));
+    }
+
+    public void add(DeleteRequest request) {
+        updateState(request, BulkRequest.getSize(request));
+    }
+
+    public void add(UpdateRequest request) {
+        updateState(request, BulkRequest.getSize(request));
+    }
+
+    private void updateState(DocWriteRequest<?> request, long finalSizeInBytes) {
+        if (closed.get()) {
+            throw new IllegalStateException("bulk process already closed");
+        }
+        State newState = state.updateAndGet(s -> {
+            long size = s.size + finalSizeInBytes;
+            PersistentStack<DocWriteRequest<?>> actions = s.actions.push(request);
+            if (actions.size() >= bulkActions || size >= bulkSize) {
+                return new State(PersistentStack.empty(), actions, 0);
+            } else {
+                return new State(actions, null, size);
+            }
+        });
+
+        if (newState.toFlush != null) {
+            flush(newState.toFlush);
+        }
```

Comment:
What happens if between line 79 and 89 a background thread (lets say the scheduled flush) executes the `flush()` method? Could we find ourselves calling flush with the same payload twice here?

Reply:
Wait, nevermind, I missed that the payloads are separated into different stacks. Nice.